sktime/forecasting/reconcile.py

```diff
@@ -0,0 +1,355 @@
+        X : pd.DataFrame, default=None
+            Exogenous variables based to the base forecaster
+        update_params : bool, optional (default=True)
+            whether model parameters should be updated
+
+        Returns
+        -------
+        self : reference to self
+        """
+        if y.index.nlevels < 2:
+            self.forecaster_.update(y, X, update_params=update_params)
+            return self
+
+        # bug in self.forecaster_.predict_residuals() for heir data
+        fh_resid = ForecastingHorizon(
+            y.index.get_level_values(-1).unique(), is_relative=False
+        )
+        self.residuals = y - self.forecaster_.predict(fh=fh_resid, X=X)
+
+        # could implement something specific here
+        # for now just refit
+        if self.method == "mint_cov":
+            self.g_matrix = self._get_g_matrix_mint(shrink=False)
+        elif self.method == "mint_shrink":
+            self.g_matrix = self._get_g_matrix_mint(shrink=True)
+        elif self.method == "wls_var":
+            self.g_matrix = self._get_g_matrix_mint(shrink=False, diag_only=True)
+        else:
+            raise RuntimeError("unreachable condition, error in _check_method")
+
+        return self
+
+    def _get_g_matrix_mint(self, shrink=False, diag_only=False):
+        """Define the G matrix for the MinT methods based on model residuals.
+
+        Reconciliation methods require the G matrix. The G matrix is used to redefine
+        base forecasts for the entire hierarchy to the bottom-level only before
+        summation using the S matrix.
+
+        Please refer to [1]_ for further information.
+
+        Parameters
+        ----------
+        shrink:  bool, optional (default=False)
+            Shrink the off diagonal elements of the sample covariance matrix.
+            according to the method in [2]_
+        diag_only: bool, optional (default=False)
+            Remove the off-diagonal elements of the sample covariance matrix.
+
+        Returns
+        -------
+        g_mint : pd.DataFrame with rows equal to the number of bottom level nodes
+            only, i.e. with no aggregate nodes, and columns equal to the number of
+            unique nodes in the hierarchy. The matrix indexes is inherited from the
+            input data, with the time level removed.
+
+        References
+        ----------
+        .. [1] https://otexts.com/fpp3/hierarchical.html
+        .. [2] https://doi.org/10.2202/1544-6115.1175
+        """
+        if self.residuals.index.nlevels < 2:
+            return None
+
+        # copy for further mods
+        resid = self.residuals.copy()
+        # mean scale
+        grp_range = np.arange(self.residuals.index.nlevels - 1).tolist()
+        resid = resid.groupby(level=grp_range).apply(lambda x: x - x.mean())
+        # cov matrix
+        resid = resid.unstack().transpose()
+        nobs = len(resid)
+        cov_mat = resid.transpose().dot(resid) / (nobs - 1)
```

Comment:
can't we use numpy.cov here?

Reply:
yep, I simplified this section a wee bit! 👍